application-diagram-api/src/main/java/com/xwiki/diagram/internal/handlers/DiagramMacroSolrMetadataExtractor.java

```diff
@@ -65,32 +73,81 @@ public class DiagramMacroSolrMetadataExtractor implements SolrEntityMetadataExtr
     @Inject
     private EntityNameValidationConfiguration entityNameValidationConfiguration;
 
+    @Inject
+    private Logger logger;
+
     @Override
     public boolean extract(XWikiDocument document, SolrInputDocument solrDocument)
     {
 
-        List<Block> macroBlocks = document.getXDOM().getBlocks(new MacroBlockMatcher("diagram"), Block.Axes.CHILD);
-        if (macroBlocks != null && macroBlocks.size() > 0) {
+        XDOM xdom = document.getXDOM();
+        List<Block> macroBlocks = xdom.getBlocks(new MacroBlockMatcher("diagram"), Block.Axes.CHILD);
+        if (macroBlocks != null && macroBlocks.size() > 0 && !updateMacroReference(document, xdom, macroBlocks)) {
             List<DocumentReference> macroReferences = new ArrayList<>();
             for (Block macroBlock : macroBlocks) {
-                String referenceName =
-                    macroBlock.getParameter(REFERENCE) != null ? macroBlock.getParameter(REFERENCE) : "Diagram";
-
-                String transformedNamed = this.transformName(referenceName);
 
                 DocumentReference macroReference =
-                    explicitDocumentReferenceResolver.resolve(transformedNamed, document.getDocumentReference());
+                    explicitDocumentReferenceResolver.resolve(macroBlock.getParameter(REFERENCE),
+                        document.getDocumentReference());
                 macroReferences.add(macroReference);
             }
             return linkRegistry.registerBacklinks(solrDocument, macroReferences);
+        }
+        return false;
+    }
 
+    /**
+     * Checks and updates the references of all the diagram macro calls to make sure that all of them respect the
+     * current name strategy.
+     *
+     * @param document document with all the macro calls
+     * @param xdom of the @document
+     * @param macroBlocks list of all the diagram macro calls
+     * @return true if all the references are valid, false otherwise.
+     */
+    private boolean updateMacroReference(XWikiDocument document, XDOM xdom, List<Block> macroBlocks)
+    {
+        try {
 
+            XWikiContext context = contextProvider.get();
+            boolean modified = false;
+            for (Block macroBlock : macroBlocks) {
+                String referenceName =
+                    macroBlock.getParameter(REFERENCE) != null ? macroBlock.getParameter(REFERENCE) : "Diagram";
+                // For backwards compatibility we check if the page already exists so we won't modify it.
+                DocumentReference macroReference =
+                    explicitDocumentReferenceResolver.resolve(referenceName, document.getDocumentReference());
+                boolean diagramAlreadyExists = context.getWiki().exists(macroReference, context);
```

Comment:
What happens if the document exists and the reference is not ok? We still display it and let the name handling to the diagram itself, right? A comment would be good

Reply:
If the document exists and the reference doesn't respect the naming strategy the macro will be skipped and keep the current reference to breaking the macro calls that already exist.